File: leetcode/494/TargetSum.js
/**
 * @param {number[]} nums
 * @param {number} S
 * @return {number}
 */
// O(2^n)
// var findTargetSumWays = function(nums, S) {
//     let total = 0;

//     function find(index, t) {
//         if (index >= nums.length) {
//             if (t === S) total++;
//             return;
//         }

//         find(index + 1, t - nums[index]);
//         find(index + 1, t + nums[index]);
//     }

//     find(0, 0);

//     return total;
// };


/**
 * @param {number[]} nums
 * @param {number} S
 * @return {number}
 */
var findTargetSumWays = function(nums, S) {
    let count = 0;
    
    function calculate(i, sum) {
        if (i >= nums.length) {
            if (sum === S)
                count++;
        }else
        {
            calculate(i + 1, sum +nums[i]);
            calculate(i + 1, sum -nums[i]);
        }
        
    }
    
    calculate(0, 0);
    return count;
    
};

console.log(findTargetSumWays([1,1,1,1,1],3));

File: leetcode/494/TargetSum.java
import java.util.*;

class TargetSum {
    public static void main(String[] args) {
        Solution sl = new Solution();
        int[] nums = { 1, 1, 1, 1, 1 };
        System.out.println(sl.findTargetSumWays(nums, 3));
    }
}

class Solution {
    int count = 0;

    public int findTargetSumWays(int[] nums, int S) {
        calculate(nums, 0, 0, S);
        return count;
    }

    public void calculate(int[] nums, int i, int sum, int S) {
        if (i == nums.length) {
            if (sum == S)
                count++;
        } else {
            calculate(nums, i + 1, sum + nums[i], S);
            calculate(nums, i + 1, sum - nums[i], S);
        }
    }
}

class Solution1 {
    int count = 0;

    public int findTargetSumWays(int[] nums, int S) {
        int[][] memo = new int[nums.length][2001];
        for (int[] row : memo)
            Arrays.fill(row, Integer.MIN_VALUE);
        return calculate(nums, 0, 0, S, memo);
    }

    public int calculate(int[] nums, int i, int sum, int S, int[][] memo) {
        if (i == nums.length) {
            if (sum == S)
                return 1;
            else
                return 0;
        } else {
            if (memo[i][sum + 1000] != Integer.MIN_VALUE) {
                return memo[i][sum + 1000];
            }
            int add = calculate(nums, i + 1, sum + nums[i], S, memo);
            int subtract = calculate(nums, i + 1, sum - nums[i], S, memo);
            memo[i][sum + 1000] = add + subtract;
            return memo[i][sum + 1000];
        }
    }
}